ms_active_directory/core/ad_session.py

```diff
@@ -1688,19 +1688,32 @@ def find_groups_for_entities(self, entities: List[Union[str, ADObject]], attribu
         entity_dns_to_entities = ldap_utils.normalize_entities_to_entity_dns(entities, lookup_by_name_fn, controls,
                                                                              skip_validation)
 
-        filter_pieces = []
+        filter_pieces = set()
         mapping_dict = {}
-        for entity_dn, entity in entity_dns_to_entities:
+        primary_group_dict = {}
+        for entity_dn, entity in entity_dns_to_entities.items():
             filter_piece = '({}={})'.format(ldap_constants.AD_ATTRIBUTE_MEMBER,
                                             ldap_utils.escape_dn_for_filter(entity_dn))
-            filter_pieces.append(filter_piece)
+            filter_pieces.add(filter_piece)
             mapping_dict[entity] = []
             if include_primary:
-                entity_obj = self.find_object_by_distinguished_name(entity_dn, controls=controls)
-                # only get primary group if entity is of type ADUser or ADComputer
-                if isinstance(entity_obj, ADUser) or isinstance(entity_obj, ADComputer):
-                    primary_group = self.find_primary_group_for_account(entity_obj, attributes_to_lookup, controls)
-                    mapping_dict[entity].append(primary_group)
+                # get objectSid and primaryGroupID for entity
+                attributes = [ldap_constants.AD_ATTRIBUTE_OBJECT_SID, ldap_constants.AD_ATTRIBUTE_PRIMARY_GROUP_ID]
+                entity_obj = self.find_object_by_distinguished_name(entity_dn, attributes, controls)
+
+                entity_sid = entity_obj.get(ldap_constants.AD_ATTRIBUTE_OBJECT_SID)
+                primary_group_id = entity_obj.get(ldap_constants.AD_ATTRIBUTE_PRIMARY_GROUP_ID)
+
+                # only add the filter for primary group if the entity has objectSid and primaryGroupID attributes
+                if entity_sid and primary_group_id:
+                    # get the objectSid for the primary group given the entity's objectSid and the primaryGroupID
+                    group_sid = ldap_utils.construct_primary_group_sid(entity_sid, primary_group_id)
+                    primary_filter_piece = '({}={})'.format(ldap_constants.AD_ATTRIBUTE_OBJECT_SID, group_sid)
```

Comment:
when you filter by SID, you actually need to specially format it because SIDs are bytestrings with a neat string representation.

if you look at `find_object_by_sid` I turned arbitrary formats into bytes this way
```
        sid_obj_or_bytes = sid
        if isinstance(sid, str) and sid.upper().startswith('S-'):
            logger.debug('Converting string SID to ObjectSID object')
            sd = sd_utils.ObjectSid()
            sd.from_canonical_string_format(sid.upper())
            sid_obj_or_bytes = sd
        elif isinstance(sid, security_constants.WellKnownSID):
            logger.debug('Converting Well Known SID to ObjectSID object')
            sd = sd_utils.ObjectSid()
            sd.from_canonical_string_format(sid.value)
            sid_obj_or_bytes = sd

        # we must either have an ObjectSid or bytes at this point. we want to end up with bytes
        sid_bytes = sid_obj_or_bytes
        if isinstance(sid_obj_or_bytes, sd_utils.ObjectSid):
            sid_bytes = sid_obj_or_bytes.get_data()
```

and then escaped it using `ldap_utils.escape_bytestring_for_filter` - so you'll need to do that.

I'm also worried that primary groups are going to very swiftly push the filter length past the maximum, since sid bytestrings are fairly long, so maybe add some logging about that (you'll see that I already have some logging about that below)

Reply:
From what I've seen, according to Microsoft TechNet Wiki, you can filter objectSid either by a bytestring or the decimal display format. 

> The objectSID attribute is saved in Active Directory as a byte array. You can either specify the decimal display format S-1-5-21-73586283-152049171-839522115-1111 or the equivalent hex representation where each byte is escaped
"\01\05\00\00\00\00\00\05\15\00\00\00\6B\D6\62\04\13\16\10\09\43\17\0A\32\57\04\00\00".

> Object with SID "S-1-5-21-73586283-152049171-839522115-1111"
-> filter: (objectSID=S-1-5-21-73586283-152049171-839522115-1111)
Object with SID "0105000000000005150000006BD662041316100943170A3257040000"
-> filter: (objectSID=\01\05\00\00\00\00\00\05\15\00\00\00\6B\D6\62\04\13\16\10\09\43\17\0A\32\57\04\00\00)

[Source](https://social.technet.microsoft.com/wiki/contents/articles/5392.active-directory-ldap-syntax-filters.aspx)

Given this, for our purposes, I would think it would be better to format the filter as a string to keep the length of the overall chain filter to a minimum.

Reply:
huh, TIL